openid-zkp-auth/circuits/helpers/strings.circom

```diff
@@ -0,0 +1,192 @@
+pragma circom 2.1.3;
+
+include "../../node_modules/circomlib/circuits/comparators.circom";
+include "../../node_modules/circomlib/circuits/multiplexer.circom";
+include "../../node_modules/circomlib/circuits/gates.circom";
+include "misc.circom";
+include "base64.circom";
+
+/**
+SliceFixed: Returns a fixed-length slice of an array.
+More precisely, in[index:index+outLen] (both inclusive).
+
+Cost: inLen + outLen * inLen
+
+Range checks:
+    - index in [0, inLen)
+    - index + outLen in [0, inLen]
+    - outLen in (0, inLen]
+**/
+template SliceFixed(inLen, outLen) {
+    assert(outLen > 0);
+    assert(outLen <= inLen);
+
+    signal input in[inLen];
+    signal input index;
+    signal output out[outLen];
+
+    RangeCheck(numBits(inLen), inLen - 1)(index); // index in [0, inLen - 1]
+    RangeCheck(numBits(inLen), inLen)(index + outLen); // index + outLen in [0, inLen]
+
+    // eqs[index] = 1, 0 otherwise
+    signal eqs[inLen] <== OneBitVector(inLen)(index);
```

Comment:
Not sure I follow. To explain a little more of what's going on: `EqBitVector` gives us a vector of all zeroes except at `i = index`. Then, we do a scalar product, e.g., `out[0] = Sum_i in[i] * eqs[i]`. Instead of generating a new vector for each output index, we are reusing the same vector.